Use instanceof Stringable check instead of method_exists().

ADmad opened the pull request:
<!---

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

The best way to propose a feature is to open an issue first and discuss your ideas there before implementing them.

Always follow the [contribution guidelines](https://github.com/cakephp/cakephp/blob/master/.github/CONTRIBUTING.md) when submitting a pull request. In particular, make sure existing tests still pass, and add tests for all new behavior. When fixing a bug, you may want to add a test to verify the fix.

-->

ADmad committed:
Use instanceof Stringable check instead of method_exists().

markstory reviewed (commented):
This is a nice improvement :clap:

ADmad committed:
Make I18n\DateTime and I18n\Date explicitly implement Stringable interface.

PHP doesn't seem to automatically recognize a class as instance of Stringable if
the __toString() method is available through a trait. DateFormatTrait::__toString()
in this case.

ADmad committed:
Expliclity declare Stringable interface for classes with __toString().

This is recommended in the PHP docs.
